cpp/src/join/chuck_join.cu

```diff
@@ -0,0 +1,119 @@
+  }
+#endif
+
+  if ((JoinType::LEFT_ANTI_JOIN == join_type) && (0 == right.num_rows())) {
+    // Everything matches, just copy the proper columns from the left table
+    return std::make_unique<experimental::table>(left.select(return_columns), stream, mr);
+  }
+
+  // Only care about existence, so we'll use an unordered map (other joins need a multimap)
+  using hash_table_type = concurrent_unordered_map<cudf::size_type, bool8, RowHash, row_equality>;
+
+  // Create hash table containing all keys found in right table
+  auto right_rows_d = table_device_view::create(right.select(right_on), stream);
+  size_t const hash_table_size = compute_hash_table_size(right.num_rows());
+  RowHash hash_build{*right_rows_d};
+  row_equality equality_build{*right_rows_d, *right_rows_d};
+
+  // Going to join it with left table
+  auto left_rows_d = table_device_view::create(left.select(left_on), stream);
+  RowHash hash_probe{*left_rows_d};
+  row_equality equality_probe{*left_rows_d, *right_rows_d};
+
+  auto hash_table_ptr = hash_table_type::create(hash_table_size,
+                                                std::numeric_limits<bool8>::max(),
+                                                std::numeric_limits<cudf::size_type>::max(),
+                                                hash_build,
+                                                equality_build,
+                                                hash_probe,
+                                                equality_probe);
+  auto hash_table = *hash_table_ptr;
+
+  constexpr int block_size{DEFAULT_CUDA_BLOCK_SIZE};
+  experimental::detail::grid_1d config(right.num_rows(), block_size);
+
+  build_hash_table<<<config.num_blocks, config.num_threads_per_block, 0, stream>>>(hash_table, right.num_rows());
+
+  //
+  // Now we have a hash table, we need to iterate over the rows of the left table
+  // and check to see if they are contained in the hash table
+  //
+
+  // For semi join we want contains to be true, for anti join we want contains to be false
+  bool join_type_boolean = (join_type == JoinType::LEFT_SEMI_JOIN);
+
+  rmm::device_vector<size_type> gather_map(left.num_rows());
+
+  auto gather_map_end = thrust::copy_if(rmm::exec_policy(stream)->on(stream),
+                                        thrust::make_counting_iterator<size_type>(0),
+                                        thrust::make_counting_iterator<size_type>(left.num_rows()),
+                                        gather_map.begin(),
+                                        [hash_table, join_type_boolean]
+                                        __device__ (size_type idx) {
+                                          auto pos = hash_table.find(idx);
+                                          return (pos != hash_table.end()) == join_type_boolean;
+                                        });
+
+  gather_map.resize(thrust::distance(gather_map.begin(), gather_map_end));
```

Comment:
Is this resize necessary? The gather_map will get freed when this function returns, and the iterators bound the range that will be read, so I think resizing it is unnecessary overhead?

Reply:
Resizing is not necessary.  @jrhemstad and I had a discussion on another PR about this.  In that case he was in favor of resizing in case some future maintainer wanted to use the contents of the vector for a different purpose (element from gather_map_end to gather_map.end() are undefined, so if someone were to try to run an algorithm that iterates over all of gather_map they would fail).

I do see in the thrust code that resize does call erase, so I suppose there is a cost to that (I had originally imagined that it only changed the size).  I can remove it if we think that's better.

Reply:
I don't think it's likely for some future maintainer to use the later part of gather_map, right? So I suggest removing the resize() and replacing it with a one-line comment that `gather_map_end` is the now the end of the valid data.